### PR TITLE
`elim` multiple targets

### LINKED ISSUE #62: extend `elim` to have multiple expressions/patterns
To support the followings, we should be able to allow multiple expression/pattern in `elim` first
```
def plus (n : Nat) (m : Nat) : Nat
| zero,  m => m
| suc n, m => suc $ plus n m
```
The above will lead us to have the following syntax & checking first!
```
def plus (n : Nat) (m : Nat) : Nat =>
elim n m
| zero,  m => m
| suc n, m => suc $ plus n m
```

From #60, keep going

### PR DESCRIPTION
**Pull request by dannypsnl**
* syntax
    * multi-expressions as `elim` target
    * multi-patterns in each `elim` case
* `eval` semantic
* elaboration process that convert pattern to its correct case
    1. var pattern: `m` on `Nat`

        For example, the following `m` is not going to be destruct, but just bind
        ```
        elim n, m
        | zero, m => m
        | suc n, m => suc $ plus n m
        ```

    2. ctor pattern: `zero`, `suc n` on `Nat`

resolve #62

Signed-off-by: Lîm Tsú-thuàn <dannypsnl@gmail.com>

### CODE REVIEW

File: src/Violet/Core.idr [hunk 1]
```diff
@@ -101,114 +101,109 @@ mutual
...
+				let tm = Elim ts cases'
+				pure (tm, !(convTys (ElimInfer $ tm) env rhs_tys))
 				where
-					maybeConv : Env -> Maybe VTy -> VTy -> App e (Maybe VTy)
-					maybeConv env Nothing t = pure $ Just t
-					maybeConv env (Just t) t' = do
+					kk : Name -> Name -> List Name -> Maybe (List VTy) -> App e (Pat, List (Name, VTy))
+					kk head x [] Nothing = pure (PVar head, [(head, VData x)])
+					kk head _ vars (Just tys) = pure (PCons head vars, vars `zip` tys)
+					kk head x _ Nothing = report $ BadConstructor head x
```
Comment (dannypsnl) [inline src/Violet/Core.idr:156]: fucked up naming, to fix

File: example/test.vt [hunk 2]
```diff
@@ -1,8 +1,8 @@
 module Test
 
 data Bool
-| True
-| False
+| true
+| false
```
Comment (rayshih) [inline example/test.vt:5]: Just curious, why this change?

Comment (dannypsnl) [reply example/test.vt:5]: Just a misc for no reason. Do you think we need to have convention for our constructor naming?

Comment (rayshih) [reply example/test.vt:5]: I haven't formed any opinion yet

Comment (dannypsnl) [reply example/test.vt:5]: I think we can postpone it till we start developing standard library.

File: src/Violet/Core.idr [hunk 3]
```diff
@@ -101,114 +101,109 @@ mutual
...
+						pure ((pats', rhs) :: cases, ty :: ts))
+					([], []) cases
+				let tm = Elim ts cases'
+				pure (tm, !(convTys (ElimInfer $ tm) env rhs_tys))
 				where
-					maybeConv : Env -> Maybe VTy -> VTy -> App e (Maybe VTy)
-					maybeConv env Nothing t = pure $ Just t
-					maybeConv env (Just t) t' = do
+					elabPattern : Name -> Name -> List Name -> Maybe (List VTy) -> App e (Pat, List (Name, VTy))
+					elabPattern head x [] Nothing = pure (PVar head, [(head, VData x)])
```
Comment (rayshih) [inline src/Violet/Core.idr:154]: So this suggests if the constructor takes no argument, then there won't be record found by using `findCtorSet`, right? What is the downside of also adding 0-arg constructor into the mapping?

Comment (dannypsnl) [reply src/Violet/Core.idr:154]: No, this is “if no vars spine, and can’t find in ctor set, then it’s a var pattern”

A 0-arg constructor will go to next case

Comment (rayshih) [reply src/Violet/Core.idr:154]: I mean why can't they be the same case?

If we make constructors with "no vars spine" also in ctor set, then we won't need to have 2 separate matches (line 154, 155) and we can probably get rid of pattern matching in line 164

File: src/Violet/Core.idr [hunk 4]
```diff
@@ -101,114 +101,109 @@ mutual
...
+			go (Elim ts cases) = do
+				tys <- for ts (infer env ctx)
+				(cases', rhs_tys) <- foldlM
+					(\(cases, ts), elim_case => do
+						let (_, rhs) = elim_case
+						(pats', ty) <- checkCase env ctx (map snd tys) elim_case
+						pure ((pats', rhs) :: cases, ty :: ts))
+					([], []) cases
+				let tm = Elim ts cases'
+				pure (tm, !(convTys (ElimInfer $ tm) env rhs_tys))
```
Comment (rayshih) [inline src/Violet/Core.idr:151]: extra dollar sign?

File: src/Violet/Core/Val.idr [hunk 5]
```diff
@@ -96,19 +94,33 @@ eval env tm = case tm of
...
+					then pure $ (names `zip` vals) ++ !(matches next next')
+					else Nothing
+			matches [] [] = pure []
+			matches _ _  = Nothing
+
+			go : List Val -> List (List Pat, Tm) -> Either EvalError Val
+			go vals ((pats, rhs) :: nextPat) = case matches pats vals of
+				Just lenv => eval ({ local := lenv ++ env.local } env) rhs
+				Nothing => go vals nextPat
+			go vals [] = Left OutOfCase
```
Comment (rayshih) [inline src/Violet/Core/Val.idr:123]: I wonder if we want to and how to implement exhaustive check

**Review by rayshih — APPROVED** [review]
thanks for the patience. I learnt a lot

File: src/Violet/Core.idr [hunk 6]
```diff
@@ -101,114 +101,109 @@ mutual
...
+				Right b' <- pure $ b u'
+					| Left e => report $ cast e
+				pure (Apply t u, b')
 			go (Lam {}) = report (InferLam tm)
 			go (Pi x a b) = do
-				check env ctx a VU
-				check (extendEnv env x (VVar x)) (extendCtx ctx x !(runEval eval env a)) b VU
-				pure VU
+				a <- check env ctx a VU
+				_ <- check (extendEnv env x (VVar x)) (extendCtx ctx x !(runEval eval env a)) b VU
```
Comment (dannypsnl) [inline src/Violet/Core.idr:133]: Missing b?
The semantic is elab and check

File: src/Violet/Core.idr (hunk 4 above)
Comment (dannypsnl) [reply src/Violet/Core.idr:151]: yes, thanks!

File: src/Violet/Core/Val.idr (hunk 5 above)
Comment (dannypsnl) [reply src/Violet/Core/Val.idr:123]: That would be the task of elaborator, which is located at `infer` in the case.

For example, a indexed type has ctor set `{ a, b }`. Considering non-nested and single pattern, the checker can say if you have `| a => ...` and `| b => ...` is enough. But with nested pattern, this will not be that easy.

This problem should be solved with whole language view, therefore, postpone to the #72
